feat(physics): `quantity_simplify` can now optionally simplify units across dimensions

dyc3 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #21234

#### Brief description of what is fixed or changed
Added an optional flag, `across_dimensions` to the function `quantity_simplify`, which allows units to be simplified across dimensions, eg. joule/coulomb -> volt.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.units
  * `Quantity`: it is now possible to determine if a Quantity is prefixed (eg. `nanometers`) with the property `is_prefixed`.
  * `Quantity`: it is now possible to determine if a Quantity represents a physics constant (eg. `speed_of_light`) with the property `is_physical_constant` or if is an instance of the new subclass `PhysicalConstant`.
  * `quantity_simplify()` can now simplify units across dimensions using the optional flag `across_dimensions=True`.
<!-- END RELEASE NOTES -->

dyc3 commented:
I'm confused, am I supposed to update the authors/mailmap?

dyc3 committed:
feat(physics.units): `quantity_simplify` can now optionally simplify units across dimensions

Upabjojr commented on sympy/physics/units/util.py:
```diff
@@ -154,6 +158,46 @@ def quantity_simplify(expr):
+        # combine quantities of different dimensions into a single
+        # quantity that is equivalent to the original expression
+
+        unit_system = UnitSystem.get_unit_system("SI")
```
What about if you are not using SI unit systems?

dyc3 replied:
Ah, yeah I forgot to fix that

dyc3 replied:
I've added an extra argument for the unit system. However, I am uncertain on whether or not this sufficiently solves the problem because of the hard coded dict `preferred_dimension_to_unit`. 

Realistically though, I don't think we should consider it a blocker for merging because in 99.99% of cases, people are going to be using SI.

Upabjojr commented:
Simplifying quantities across dimensions may be dependent on the unit system that is being used (i.e. there could be different results depending on the unit system being used).

I believe the unit system needs to be specified, otherwise the results may not correspond to the desired ones.

Upabjojr commented on sympy/physics/units/util.py:
```diff
@@ -120,21 +123,24 @@ def get_total_scale_factor(expr):
 
 
-def quantity_simplify(expr):
+def quantity_simplify(expr, across_dimensions: bool=False, unit_system="SI"):
```
I would suggest to go as far as not assuming a default unit system. That is, requiring the users to specify the unit system they want to use. The purpose of the _units_ module is to be useful for generic unit systems and I would attempt at distancing from _SI_ as much as possible.

Maybe `unit_system: Optional[Union[UnitSystem, str]] = None` and then raise an exception if `across_dimensions=True` and `unit_system` is `None`?

dyc3 replied:
Other functions that take `unit_system` assume a default value of `"SI"`, like `convert_to()`. I think it would be better to at least remain consistent.

Upabjojr replied:
What about uniting the parameters `across_dimensions` and `unit_system`?

Upabjojr replied:
> Other functions that take `unit_system` assume a default value of `"SI"`, like `convert_to()`

I regret that decision.

dyc3 replied:
Uniting the parameters sounds confusing from an end user's perspective. Should this feature be moved out into a separate function? What should it be called? Something like `quantity_unify()`?

Upabjojr replied:
> Uniting the parameters sounds confusing from an end user's perspective.

Probably, it was just an idea to avoid having too many parameters.

Upabjojr commented on sympy/physics/units/util.py:
```diff
@@ -154,6 +160,46 @@ def quantity_simplify(expr):
+        from . import volt, ohm, ampere, pascal, farad, second, watt, henry, meter, newton, weber, siemens, coulomb, tesla, candela, steradian
+        from . import voltage, impedance, current, pressure, capacitance, time, power, inductance, volume, force, luminous_intensity, magnetic_flux, conductance, charge, magnetic_density
+
+        preferred_dimension_to_unit = {
```
This dictionary is assuming the base units of _SI_. I think these values can be extracted from the `unit_system` variable. This would make the code independent of the unit system.

dyc3 replied:
I agree. I'm unsure if this is possible without adding additional metadata to `Quantity`. For example, `bar` and `pascal` are both units of pressure. How do we know which one to pick?

I'll try it out and see what happens.

Upabjojr replied:
The unit system object should have a dictionary of scale factors, IIRC. Maybe that could be used to determine the base unit?

dyc3 replied:
I tried filtering units by scale factors, and I was getting some weird results. I'll push a new commit to show my current solution momentarily.

oscargus commented:
> I'm confused, am I supposed to update the authors/mailmap?

Yes, the mailmap-file, see https://github.com/sympy/sympy/wiki/Development-workflow#add-your-name-and-email-address-to-the-mailmap-file  (Authors will be updated automatically on release.)

dyc3 committed:
make `quantity_simplify()` unit system agnostic

oscargus commented:
These are the current errors in code-quality (in case you missed them):

```
 sympy/physics/units/unitsystem.py:13:1: F401 'sympy.physics.units.prefixes.BIN_PREFIXES' imported but unused
sympy/physics/units/util.py:6:1: F401 'typing' imported but unused
```

Upabjojr commented on sympy/physics/units/util.py:
```diff
@@ -154,6 +158,45 @@ def quantity_simplify(expr):
+
+        # HACK: add missing/wrong units to the map.
+        if unit_system.name == "SI":
+            preferred_dimension_to_unit.update({
```
what about this:

```python
from sympy.physics.units import definitions

preferred_dimension_to_unit = {}

for name in dir(definitions):
    quantity = getattr(definitions, name)
    if not isinstance(quantity, Quantity):
        continue
    dim = unit_system.get_quantity_dimension(quantity)
    cc = convert_to(quantity, unit_system._base_units)
    if cc.as_coeff_Mul()[0] == 1:
        preferred_dimension_to_unit[dim.name] = quantity
```

dyc3 replied:
Unfortunately, this method appears to be more correct, but it absolutely kills performance. The unit test takes noticeably longer. Also, the velocity test case fails using this method. I think we can make something like this work, if we compute it beforehand. 

It's really annoying that there are all these prefixed units like `picojoule` and `teranewton`. IMO, they should not exist, we already have the prefixes defined using `Prefix`, that should be sufficient.

dyc3 replied:
Ultimately, I think that this is going to need some metadata changes, and _maybe_ breaking changes.

Upabjojr replied:
What about using this method as a code generator? That dictionary needs to be computed only once in the end... Maybe when you install SymPy or when you use it the first time?

dyc3 replied:
I think that doing code gen is going to be harder than it sounds. An LRU cache seems like a better idea.

Upabjojr replied:
> If an application using sympy were to allow for unescaped user input, it could be possible to end up saving and executing malicious code.

Well, the code generator should not be called by end users, but by the developers before releasing a new version of SymPy.

> Regardless, I've already implemented 2 new properties for Quantity: `is_prefixed` and `is_physics_constant`. With these, I'm pretty sure the call to `convert_to()` will not be necessary, and none of this will be required. I'll commit a new implementation shortly.

That's great to hear :)

dyc3 committed:
feat(physics.units): add `is_prefixed` property to `Quantity`

dyc3 committed:
feat(physics.units): add `is_physics_constant` property to `Quantity`

dyc3 committed:
fix(physics.unit): make `prefix_unit()` create Quantities that are correctly marked as prefixed

dyc3 committed:
refactor(physics.units): rewrite `UnitSystem.get_units_non_prefixed()`

dyc3 committed:
fix(physics.unit): make sure that all base units are included in the SI unit system

dyc3 committed:
refactor(physics.unit): `quantity_simplify()` now uses `_build_preferred_dimension_to_unit()`

dyc3 committed:
fix(physics.units): derive velocity preferred units as length / time

Upabjojr commented on sympy/physics/units/quantities.py:
```diff
@@ -63,13 +64,19 @@ def __new__(cls, name, abbrev=None, dimension=None, scale_factor=None,
         obj._ascii_repr = pretty_ascii_repr
         obj._mathml_repr = mathml_presentation_repr
+        obj._is_prefixed = is_prefixed
+        obj._is_physics_constant = is_physics_constant
```
Be careful that this information is lost in case you call `Quantity(*quantity_variable.args)` or `Quantity.func(*quantity_variable.args)`, because `is_prefixed` and `is_physics_constant` are not passed to the superclass as arguments.

This could be an issue if you call some SymPy functionality like the term rewriter or the substitution.

Upabjojr replied:
What about subclassing `Quantity` into a new class `Constant` just defining `is_constant` as a static field?
e.g.
```python
class Quantity:
    is_physics_constant = False
    ...

class Constant(Quantity):
    is_physics_constant = True
```

dyc3 replied:
> Be careful that this information is lost

Could you give me a test case for this? I'm not familiar enough with this SymPy behavior.

> What about subclassing `Quantity`

That seems like a better idea.

Upabjojr commented on sympy/physics/units/tests/test_quantities.py:
```diff
@@ -510,3 +512,32 @@ def test_issue_22164():
     # deprecation warning is not expected here
     SI._collect_factor_and_dimension(expr)
+
+def test_prefixed_flag():
```
Maybe `test_prefixed_property`, I wouldn't call it a `flag` as it's supposed to be a constant.

dyc3 committed:
refactor(physics.units): add `PhysicalConstant(Quantity)` for physical constants

Upabjojr commented on sympy/physics/units/quantities.py:
```diff
@@ -232,3 +241,17 @@ def convert_to(self, other, unit_system="SI"):
+class PhysicalConstant(Quantity):
+    """Represents a physical constant, eg. `speed_of_light` or `avogadro_constant`."""
+
+    is_commutative = True
```
aren't these already inherited from the superclass?

dyc3 replied:
Ah yeah, they are. Fixed.

dyc3 commented on sympy/physics/units/systems/si.py:
```diff
@@ -56,6 +56,7 @@
 for u in units:
     all_units.extend(prefix_unit(u, PREFIXES))
 
+all_units.extend(units)
```
No, it's not because the output of `prefix_unit()` does not contain `u` without a prefix. It's very weird.

dyc3 commented on sympy/physics/units/util.py:
```diff
@@ -120,21 +123,43 @@ def get_total_scale_factor(expr):
+
+    preferred_dimension_to_unit = {}
+
+    for quantity in unit_system.get_units_non_prefixed():
```
You bring up a good point. The purpose of this is to build the lookup table using base units, because otherwise it'll end up picking units like `picometer` or `picojoule`. As a result, `gram` will be picked over `kilogram`, because `kilogram` is prefixed.

I think it would be pretty feasible to grab the degree of the resulting number, and select the closest prefix. However, I feel like it could lead to unexpected behavior. It's extra complexity to debug later, and it shouldn't go in this function. In the interest of trying to keep this PR small, I was going to follow up this PR with another to add a new function or optional arg to this function that does this.

Upabjojr replied:
Maybe the preferred units for a given unit system should just be hard-coded into a dictionary in the end?

dyc3 replied:
But then it won't be unit system agnostic.

dyc3 commented on sympy/physics/units/tests/test_quantities.py:
```diff
@@ -510,3 +512,33 @@ def test_issue_22164():
+            continue
+        if name.endswith('_constant'):
+            assert isinstance(quantity, PhysicalConstant), f"{quantity} must be PhysicalConstant, but is {type(quantity)}"
+            assert quantity.is_physics_constant, f"{name} is not marked as physics constant when it should be"
```
Yes. It helps enforce that any new constants that get added will get added correctly, using `PhysicalConstant`. If you are worried about unit test performance, there are many more egregious tests than this one.

dyc3 committed:
refactor(physics.units): rename `is_physics_constant` to `is_physical_constant`

Upabjojr commented:
By the way, why are you modifying the authors fail? Usually no one does it in a PR.

dyc3 commented:
I'm not? I only updated the mailmap file, as requested above. Idk why test/authors is still failing.

oscarbenjamin commented on .mailmap:
```diff
@@ -122,8 +122,8 @@ Alan Bromborsky <abrombo@verizon.net> <brombo@GA.(none)>
 Alex Argunov <sajkoooo@gmail.com>
-Alex Lindsay <adlinds3@ncsu.edu> <lindsayad@users.noreply.github.com>
 Alex Lindsay <adlinds3@ncsu.edu> Alex Lindsay <alexlindsay239@gmail.com>
+Alex Lindsay <adlinds3@ncsu.edu> <lindsayad@users.noreply.github.com>
```
I'm not sure why these kinds of changes sometimes happen. I guess you've used the `mailmap_update.py` script rather than the `mailmap_check.py` script but I can't be sure. Do you know which script you ran?

The authors check is failing and says this:
```
This author is not included in the .mailmap file:
Carson McManus <carson.mcmanus1@gmail.com>
This author is not included in the .mailmap file:
Carson McManus <dyc3@users.noreply.github.com>

The .mailmap file needs to be updated because there are commits with
unrecognised author/email metadata.
The mailmap file was reordered

For instructions on updating the .mailmap file see:
https://github.com/sympy/sympy/wiki/Development-workflow#add-your-name-and-email-address-to-the-mailmap-file
```
The lines that you should add to the .mailmap file would be something like:
```
Carson McManus <carson.mcmanus1@gmail.com> Carson McManus <dyc3@users.noreply.github.com>
```
That line means that `Carson McManus` with the gmail address is a listed author and that any commits with the `dyc3` email address should be associated with the same author.

After adding that line you should run the `mailmap_check.py` script twice: first time reorders the file and the second time should confirm that the file is now up to date.

dyc3 replied:
Ah yeah I did use the `mailmap_update.py` script. Seems like this process needs some better instructions haha

Upabjojr commented:
> I'm not? I only updated the mailmap file, as requested above. Idk why test/authors is still failing.

I think that script is usually run only 2 or 3 times per year, before the release of a new SymPy versions.

Upabjojr commented:
In reality the issue of the difference between physical constants and units is a lot more subtle. You can define a unit system with physical constants as units, so maybe it does not make much sense to have different classes for `Quantity` and `PhysicalConstant`. In the end, everything is just a `Quantity`... it's just a convention to use some as units and some others as physical constants. Maybe the class `PhysicalConstant` does not make that much sense.

I'm not sure how the issue of simplification across dimensions could be resolved, maybe with a dictionary of preferred units associated to the unit system? Maybe using a complexity score for the expression as used by SymPy's `simplify( )`?

dyc3 commented:
I'm a little confused. I'm not entirely sure what is blocking this, (besides the obvious merge conflicts and mailmap stuff). I think that this implementation is pretty good, and I really don't want this PR to sit and collect dust.

Upabjojr commented:
> I'm a little confused. I'm not entirely sure what is blocking this, (besides the obvious merge conflicts and mailmap stuff). I think that this implementation is pretty good, and I really don't want this PR to sit and collect dust.

I think that units and physical constants should be defined inside the `UnitSystem` class, not globally. That is the problem I see with the current implementation.

dyc3 committed:
refactor(physics.units): hard code preferred units map into each unit system

I also took the liberty of renaming some imports to be more explicit and clear.

dyc3 committed:
update mailmap

dyc3 commented:
Rebased and added preferred units as hard coded dict, controlled by the `UnitSystem` class.

I also took the liberty of renaming some imports to be more explicit and clear in some of the unit system definitions.

Upabjojr commented on sympy/physics/units/systems/mks.py:
```diff
@@ -15,19 +15,30 @@
 
 # unit system
-MKS = UnitSystem(base_units=(m, kg, s), units=all_units, name="MKS", dimension_system=dimsys_length_weight_time)
+MKS = UnitSystem(base_units=(meter, kilogram, second), units=all_units, name="MKS", dimension_system=dimsys_length_weight_time, preferred_units={
```
what about calling the dict _derived units_ instead of _preferred units_? Indeed, they are the [officially derived units](https://en.wikipedia.org/wiki/SI_derived_unit).

dyc3 replied:
Sounds good, will do.

Upabjojr commented:
One doubt: are physical constants really physical constants in all unit systems or do they depend on the chosen unit system?

For example, the speed of light is equal to 1 and dimensionless in natural units. Can it still be considered a constant?

What about a unit system defined using Planck units as base units? A lot of physical constants would turn out to be units instead of constants.

dyc3 commented:
According to the [Physical Constant](https://en.wikipedia.org/wiki/Physical_constant) article on wikipedia:

> A physical constant, sometimes fundamental physical constant or universal constant, is a [physical quantity](https://en.wikipedia.org/wiki/Physical_quantity) that is generally believed to be both universal in nature and have [constant](https://en.wikipedia.org/wiki/Constant_(mathematics)) value in time. It is contrasted with a [mathematical constant](https://en.wikipedia.org/wiki/Mathematical_constant), which has a fixed numerical value, but does not directly involve any physical measurement.

Under this definition, IMO, the speed of light is correctly defined as a fundamental physical constant, and should be denoted as such. 

Looking at our current definition for the natural unit system, I'm not completely sure if it's correct. I've filed #23261 as it is out of the scope of this PR.

dyc3 commented:
Looking for one last review, and it should be ready to merge in.

Upabjojr commented on sympy/physics/units/unitsystem.py:
```diff
@@ -192,3 +199,9 @@ def _collect_factor_and_dimension(self, expr):
         else:
             return expr, Dimension(1)
+
+    def get_units_non_prefixed(self) -> tSet[Quantity]:
```
Maybe this could be private, i.e. start with an underscore `_` character. Do end-users usually need to access this method?

Upabjojr replied:
OK.

Upabjojr commented on sympy/physics/units/unitsystem.py:
```diff
@@ -26,7 +28,7 @@ class UnitSystem(_QuantityMapper):
     _unit_systems = {}  # type: tDict[str, UnitSystem]
 
-    def __init__(self, base_units, units=(), name="", descr="", dimension_system=None):
+    def __init__(self, base_units, units=(), name="", descr="", dimension_system=None, derived_units: tDict[Symbol, Quantity]={}):
```
What about mapping `tDict[Dimension, Quantity]` instead of the name of the dimension?

dyc3 replied:
That might work, I'm not sure if `Dimension` can be used as a dictionary key.

Upabjojr commented:
OK, just one doubt about the implicit unit system. `quantity_simplify` now implicitly assumes SI units, this is dangerous. I would make the `unit_system` parameter mandatory in case `across_dimensions=True`. I would suggest to specify `unit_system=None` and then check if it is specified when `across_dimensions=True`. The purpose of the units module is to be independent of the chosen unit system, although some historical code is still implicitly assuming SI as the preferred unit system.

The problem is that once a new SymPy version will be released, we are not supposed to change the API anymore (unless you go to some lengthy deprecation cycle). It's therefore convenient to take some time and reason about a good API before it is too late to change it.

dyc3 committed:
refactor(physics.units): require `unit_system` if `across_dimensions` is true

dyc3 committed:
refactor(physics.units): change `derived_units` into type `Dict[Dimension, Quantity]`

Unfortunately, this requires another dictionary lookup, but it makes the unit system definitions make more sense.

Upabjojr commented:
I have merged this PR as it is now good enough to be included. 

Thanks for your contribution @dyc3 !
